File: c/src/cocktail_shaker_sort.c
#include "../sorting_algorithms.h"

void cocktail_shaker_sort(int *arr, size_t n) {
    if (n == 0) return;
    size_t start = 0;
    size_t end = n - 1;
    int swapped = 1;
    while (swapped) {
        swapped = 0;
        for (size_t i = start; i < end; ++i) {
            if (arr[i] > arr[i + 1]) {
                int tmp = arr[i]; arr[i] = arr[i + 1]; arr[i + 1] = tmp;
                swapped = 1;
            }
        }
        if (!swapped) break;
        swapped = 0;
        if (end > 0) end--;
        for (size_t i = end; i > start; --i) {
            if (arr[i - 1] > arr[i]) {
                int tmp = arr[i - 1]; arr[i - 1] = arr[i]; arr[i] = tmp;
                swapped = 1;
            }
        }
        start++;
    }
}

File: python/sorting_algorithms/algos/cocktail_shaker_sort.py
from typing import Sequence, Optional
from ..base import _setup_tracer, _finalize, SortTracer, Number


def cocktail_shaker_sort(data: Sequence[Number], *, trace: bool = False, tracer: Optional[SortTracer] = None):
    arr = list(data)
    tracer, ret_trace = _setup_tracer("cocktail_shaker_sort", trace, tracer)
    n = len(arr)
    start, end = 0, n - 1
    swapped = True
    while swapped:
        swapped = False
        for i in range(start, end):
            if tracer:
                tracer.record(arr, (i, i + 1), "forward compare")
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True
                if tracer:
                    tracer.record(arr, (i, i + 1), "swap")
        if not swapped:
            break
        swapped = False
        end -= 1
        for i in range(end - 1, start - 1, -1):
            if tracer:
                tracer.record(arr, (i, i + 1), "backward compare")
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True
                if tracer:
                    tracer.record(arr, (i, i + 1), "swap")
        start += 1
    return _finalize(arr, tracer, ret_trace)
